In [1]:
import pandas as pd

# Load the Titanic dataset
url = "https://web.stanford.edu/class/archive/cs/cs109/cs109.1166/stuff/titanic.csv"
titanic_data = pd.read_csv(url)

# Print the missing values for each column
missing_values = titanic_data.isnull().sum()
print(missing_values)


Survived                   0
Pclass                     0
Name                       0
Sex                        0
Age                        0
Siblings/Spouses Aboard    0
Parents/Children Aboard    0
Fare                       0
dtype: int64


In [2]:
import pandas as pd

# Load the Titanic dataset with missing values
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
titanic_data = pd.read_csv(url)

# Print the missing values for each column
missing_values = titanic_data.isnull().sum()
print(missing_values)


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


Summary of Interaction:

https://chatgpt.com/share/ab1ed53e-41f6-4018-a378-f09c69b89991

In this session, we discussed finding an amusing, funny, or otherwise interesting dataset containing missing values that is available in CSV format. Initially, I was provided with a link to an IMDb movie dataset, but it did not meet the requirement of containing missing values. Upon request, another dataset was provided, which was the Titanic passenger dataset, but it turned out that the version linked did not have missing values either.

After further clarification, a new Titanic dataset with missing values was provided via a link from GitHub. Finally, Python code using the pandas library was shared to load this dataset, and a script was given to display the missing values for each column in the dataset.

In [1]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
data = pd.read_csv(url)

# Display the columns in the dataset
print("Columns in the dataset:")
print(data.columns)

# Display the size of the dataset
print("\nSize of the dataset (rows, columns):")
print(data.shape)

Columns in the dataset:
Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

Size of the dataset (rows, columns):
(891, 12)


Observations are rows. Each observation in the titanic dataset is, in this context, a single passenger and all of their information (PassengerID, whether they survived, the passenger's class, name, sex, age, as well as other information), so refers to a row in the dataset. 

Variables are columns. Each variable in the titanic dataset shows a different characteristic of each passenger, such as age (and lists all of the different ages of the passengers in the titanic dataset), class, sex, fare, cabin and other categories. Some of these are qualitative (name, sex), while others are quantitative (fare, passenger ID)

In [2]:
print(data)

     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ... 

In [5]:
print(data.describe())

       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200  


In [6]:
# Display value counts for a specific column (e.g., 'Sex' or 'Embarked')
print(data['Sex'].value_counts())

Sex
male      577
female    314
Name: count, dtype: int64


In [7]:
print(data['Embarked'].value_counts())

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64


In [8]:
print(data['Survived'].valuecounts())

AttributeError: 'Series' object has no attribute 'valuecounts'

In [9]:
print(data.describe('Survived'))

TypeError: '<=' not supported between instances of 'int' and 'numpy.str_'

In [10]:
data.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [11]:
data['Survived']

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

In [12]:
print(data['Survived'].isnull())

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Survived, Length: 891, dtype: bool


In [13]:
print(data['Survived'].isnull().sum())

0


In [14]:
print(data['Survived'].value_counts())

Survived
0    549
1    342
Name: count, dtype: int64


In [15]:
data.describe('Name')

TypeError: '<=' not supported between instances of 'int' and 'numpy.str_'

In [16]:
print(data.describe())

       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200  


In [17]:
print(data.describe('PassengerId'))

TypeError: '<=' not supported between instances of 'int' and 'numpy.str_'

In [18]:
print(data['PassengerId'].describe())

count    891.000000
mean     446.000000
std      257.353842
min        1.000000
25%      223.500000
50%      446.000000
75%      668.500000
max      891.000000
Name: PassengerId, dtype: float64


In [19]:
print(data['Age'].describe())

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64


In [20]:
print(data['Pclass'].describe())

count    891.000000
mean       2.308642
std        0.836071
min        1.000000
25%        2.000000
50%        3.000000
75%        3.000000
max        3.000000
Name: Pclass, dtype: float64


In [21]:
print(data['Age'].isnull())

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888     True
889    False
890    False
Name: Age, Length: 891, dtype: bool


In [22]:
print(data['Age'].isnull().sum())

177


In [23]:
print(data.isnull().sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


In [24]:
# Example: Access rows where 'Age' has missing values
missing_age = data[data['Age'].isnull()]
print(missing_age)

     PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Tic

In [27]:
data_1 = [data['Name'] == "Moran"]

In [28]:
print(data_1)

[0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Name, Length: 891, dtype: bool]


In [29]:
data_1 = [data['Name'] == "Moran, Mr. James"]

In [30]:
print(data_1)

[0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Name, Length: 891, dtype: bool]


In [31]:
specific_name = "Allen, Miss. Elisabeth Walton"
name_match = data[data['Name'] == specific_name]
print(name_match)

     PassengerId  Survived  Pclass                           Name     Sex  \
730          731         1       1  Allen, Miss. Elisabeth Walton  female   

      Age  SibSp  Parch Ticket      Fare Cabin Embarked  
730  29.0      0      0  24160  211.3375    B5        S  


In [32]:
print(data_1 = [data['Name'] == "Moran, Mr. James"], if case=True

SyntaxError: invalid syntax (192336141.py, line 1)

In [33]:
data[data['Name'] == "Moran, Mr. James"]

PassengerId  Survived  Pclass              Name   Sex  Age  SibSp  Parch  \
5            6         0       3  Moran, Mr. James  male  NaN      0      0   

   Ticket    Fare Cabin Embarked  
5  330877  8.4583   NaN        Q

In [34]:
data[data['Pclass'] == 1]

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
11            12         1       1   
23            24         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
23                        Sloper, Mr. William Thompson    male  28.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
11       0    113783  26.5500         C103        S  
23       0    113788  35.5000           A6        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[216 rows x 12 columns]

In [35]:
data[data['Pclass'] == 1].groupby('Embarked')

In [36]:
print(data[data['Pclass'] == 1].groupby('Embarked').size())

Embarked
C     85
Q      2
S    127
dtype: int64


In [37]:
print(data[data['Pclass'] == 1].groupby('Embarked'))

In [40]:
print(data[data['Pclass'] == 1].groupby('Embarked').count())

          PassengerId  Survived  Pclass  Name  Sex  Age  SibSp  Parch  Ticket  \
Embarked                                                                        
C                  85        85      85    85   85   74     85     85      85   
Q                   2         2       2     2    2    2      2      2       2   
S                 127       127     127   127  127  108    127    127     127   

          Fare  Cabin  
Embarked               
C           85     66  
Q            2      2  
S          127    106  


In [41]:
print(data[data['Pclass'] == 3].groupby('Embarked').size())

Embarked
C     66
Q     72
S    353
dtype: int64


In [42]:
print(data[data['Pclass'] == 1].groupby('Embarked').size())

Embarked
C     85
Q      2
S    127
dtype: int64


As we can see, the total number of first class passengers is much lower than the total number of third class passengers. In both cases, the most popular port people left from was S (Southampton). The only two variables with missing values are 'Age' and 'Cabin'. 

In [43]:
corrected_age = data[data['Age'].dropna()]

KeyError: "None of [Index([22.0, 38.0, 26.0, 35.0, 35.0, 54.0,  2.0, 27.0, 14.0,  4.0,\n       ...\n       25.0, 33.0, 22.0, 28.0, 25.0, 39.0, 27.0, 19.0, 26.0, 32.0],\n      dtype='float64', length=714)] are in the [columns]"

In [44]:
corrected_age = data[data['Age'].notna()]


In [45]:
print(corrected_age)

     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ... 

In [46]:
corrected_age.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   714.000000  714.000000  714.000000  714.000000  714.000000   
mean    448.582633    0.406162    2.236695   29.699118    0.512605   
std     259.119524    0.491460    0.838250   14.526497    0.929783   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     222.250000    0.000000    1.000000   20.125000    0.000000   
50%     445.000000    0.000000    2.000000   28.000000    0.000000   
75%     677.750000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    5.000000   

            Parch        Fare  
count  714.000000  714.000000  
mean     0.431373   34.694514  
std      0.853289   52.918930  
min      0.000000    0.000000  
25%      0.000000    8.050000  
50%      0.000000   15.741700  
75%      1.000000   33.375000  
max      6.000000  512.329200

In [47]:
# Check for non-numeric columns
non_numeric_columns = data.select_dtypes(exclude=['number']).columns
print("Non-numeric columns:", non_numeric_columns)


Non-numeric columns: Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')


In [48]:
print(data.select_dtypes('object').columns)

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')


In [55]:
print(data['Age'].isnull().sum())

177


In [56]:
print(data['Cabin'].isnull().sum())

687


These show the missing values I have found in my dataset. As you can see above, I have also corrected the 'Age' column and have a 'corrected_age' variable where there are no missing values. 

In [57]:
data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [58]:
data.shape()

TypeError: 'tuple' object is not callable

In [59]:
print(data.shape)

(891, 12)


'count' is the number of values
'mean' is the average value in the column
'std' is the standard deviation, or the spread of the results
'min' is the minimum value
'25%' is the first quartile value 
'50%' is the median
'75%' is the upper quartile value
'max' is the maximum value


Only 7 columns/variables are included in data.describe() because there are the ones with numeric values, so can be used in calculations. Age has a different count than 891 (which is what data.shape gave us) due to 177 missing values as seen above. The 'Cabin' category also has missing values but is not included in the describe function because it is a non-numeric value.

5. data.shape is an attribute and does not perform any specific operations when it is accessed, it just repeats and prints the data it has stored in its memory, which is why it does not use parentheses (). It is static because its output does not change unless there is a change in the dataframe. However, data.describe() uses parentheses because it is a method function and recalculates summary statistics whenever the function is used. 
data.describe also automatically removes missing values when performing summary statistics as seen above, as there was no difference in the summary statistics for the 'Age' column (which had missing values) when we used the original dataset, and when we calcualted the 'corrected_age' summary statistics. 

This is my chat history: https://chatgpt.com/share/ab1ed53e-41f6-4018-a378-f09c69b89991

I will now be using a new chat for the rest of my questions :)

7. df.dropna() is more efficient when it is used for larger datasets with a few scattered missing values that are not all in the same column, but are in many different columns, as it will delete the specific rows, which should be a small percentage in such a large dataset, without significantly altering the entire dataset and any further analysis. 

del df['col'] is more problematic in this situation, as if there are many columns with missing values (even if each column only has 1-2 missing values), entire columns and multiple valid and valuable entries are deleted and this reduces the sample size which is used for calculations. 

del df['col'] is better when most missing values are concentrated in one single or two columns, as these columns are likely to not be very useful if they largely consist of missing values, and will also not lead to an erosion of other important entries which may have complete information in the other columns. 

A most efficient way would be to use del df['col'] as this would remove the majority of missing values that are concentrated in a few columns, leaving the dataset with limited, few scattered missing values across multiple columns. Using df.dropna() on these missing values is now much better as now only a few rows are deleted, whereas if df.dropna() was used from the beginning, many more useful entries would have been deleted. 

In [1]:
del df['Cabin']

NameError: name 'df' is not defined

9. Yes I attended office hours

In [2]:
del data['Cabin']

NameError: name 'data' is not defined

In [6]:
data

NameError: name 'data' is not defined

In [7]:
import pandas as pd

url = "https://web.stanford.edu/class/archive/cs/cs109/cs109.1166/stuff/titanic.csv"
titanic_data = pd.read_csv(url)

In [8]:
data = titanic_data

In [9]:
del data.dropna['Cabin']

TypeError: 'method' object does not support item deletion

In [10]:
del data['Cabin']

KeyError: 'Cabin'

In [11]:
data

Survived  Pclass                                               Name  \
0           0       3                             Mr. Owen Harris Braund   
1           1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2           1       3                              Miss. Laina Heikkinen   
3           1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4           0       3                            Mr. William Henry Allen   
..        ...     ...                                                ...   
882         0       2                               Rev. Juozas Montvila   
883         1       1                        Miss. Margaret Edith Graham   
884         0       3                     Miss. Catherine Helen Johnston   
885         1       1                               Mr. Karl Howell Behr   
886         0       3                                 Mr. Patrick Dooley   

        Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  
0      male  22.0                        1                        0   7.2500  
1    female  38.0                        1                        0  71.2833  
2    female  26.0                        0                        0   7.9250  
3    female  35.0                        1                        0  53.1000  
4      male  35.0                        0                        0   8.0500  
..      ...   ...                      ...                      ...      ...  
882    male  27.0                        0                        0  13.0000  
883  female  19.0                        0                        0  30.0000  
884  female   7.0                        1                        2  23.4500  
885    male  26.0                        0                        0  30.0000  
886    male  32.0                        0                        0   7.7500  

[887 rows x 8 columns]

In [12]:
import pandas as pd

# Load the Titanic dataset with missing values
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
data = pd.read_csv(url)


In [13]:
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

(Before), shows multiple entires and NaNs in the Cabin column as well as NaNs in the Age column.

In [14]:
del data['Cabin']

In [15]:
data.dropna()

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Embarked  
0        0         A/5 21171   7.2500        S  
1        0          PC 17599  71.2833        C  
2        0  STON/O2. 3101282   7.9250        S  
3        0            113803  53.1000        S  
4        0            373450   8.0500        S  
..     ...               ...      ...      ...  
885      5            382652  29.1250        Q  
886      0            211536  13.0000        S  
887      0            112053  30.0000        S  
889      0            111369  30.0000        C  
890      0            370376   7.7500        Q  

[712 rows x 11 columns]

The above shows that after deleting the column 'Cabin', which had more than 600 missing values using def data[], and then using data.dropna(), this was more efficient and retained the most amount of useful entries

In [16]:
data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [17]:
data_1 = data.dropna()

In [18]:
data_1

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Embarked  
0        0         A/5 21171   7.2500        S  
1        0          PC 17599  71.2833        C  
2        0  STON/O2. 3101282   7.9250        S  
3        0            113803  53.1000        S  
4        0            373450   8.0500        S  
..     ...               ...      ...      ...  
885      5            382652  29.1250        Q  
886      0            211536  13.0000        S  
887      0            112053  30.0000        S  
889      0            111369  30.0000        C  
890      0            370376   7.7500        Q  

[712 rows x 11 columns]

In [19]:
data_1.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   712.000000  712.000000  712.000000  712.000000  712.000000   
mean    448.589888    0.404494    2.240169   29.642093    0.514045   
std     258.683191    0.491139    0.836854   14.492933    0.930692   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     222.750000    0.000000    1.000000   20.000000    0.000000   
50%     445.000000    0.000000    2.000000   28.000000    0.000000   
75%     677.250000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    5.000000   

            Parch        Fare  
count  712.000000  712.000000  
mean     0.432584   34.567251  
std      0.854181   52.938648  
min      0.000000    0.000000  
25%      0.000000    8.050000  
50%      0.000000   15.645850  
75%      1.000000   33.000000  
max      6.000000  512.329200

After: This has retained entries and not made a large difference to the calculations.

df.groupby("col1")["col2"].describe() summarizes statistics of the column (col2) in relation to the other column (col1) that satisfy and are categorised according to this column. 
Descriptive statistics computed are: count, mean, std, min, 25%, 50%, 75%, and max of col2.

In [20]:
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Embarked  
0        0         A/5 21171   7.2500        S  
1        0          PC 17599  71.2833        C  
2        0  STON/O2. 3101282   7.9250        S  
3        0            113803  53.1000        S  
4        0            373450   8.0500        S  
..     ...               ...      ...      ...  
886      0            211536  13.0000        S  
887      0            112053  30.0000        S  
888      2        W./C. 6607  23.4500        S  
889      0            111369  30.0000        C  
890      0            370376   7.7500        Q  

[891 rows x 11 columns]

In [21]:
data_1

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Embarked  
0        0         A/5 21171   7.2500        S  
1        0          PC 17599  71.2833        C  
2        0  STON/O2. 3101282   7.9250        S  
3        0            113803  53.1000        S  
4        0            373450   8.0500        S  
..     ...               ...      ...      ...  
885      5            382652  29.1250        Q  
886      0            211536  13.0000        S  
887      0            112053  30.0000        S  
889      0            111369  30.0000        C  
890      0            370376   7.7500        Q  

[712 rows x 11 columns]

In [22]:
data_1.groupby('Pclass')['Embarked'].describe()

count unique top freq
Pclass                      
1        184      3   S  108
2        173      3   S  156
3        355      3   S  290

Handling of missing data

df.describe() gives overall summary statistics for columns and automatically excludes missing values in their calculations. 

df.groupby("col1")["col2"].describe() gives summary statistics for a specific column based on another column as long as it satisifes the conditions introduced (the two columns). It is mainly used for comparison of the distribution of data, for example to see how many first class passengers were from Southampton, as opposed ot second and third class passengers on the titanic. 

In [1]:
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
titanic_data = pd.read_csv(url)

missing_values = titanic_data.isnull().sum()
print(missing_values)

NameError: name 'pd' is not defined

ChatGPT provided the right solution, to import pandas, same results seen on Google.

In [2]:
import pandas as pd
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanics.csv"
titanic_data = pd.read_csv(url)

missing_values = titanic_data.isnull().sum()
print(missing_values)

HTTPError: HTTP Error 404: Not Found

ChatGPT said there was an error in the URl, this was less clear on Google

In [3]:
DF.groupby('Pclass')['Embarked']

NameError: name 'DF' is not defined

In [4]:
DATA.groupby('Pclass')['Embarked']

NameError: name 'DATA' is not defined

ChatGPT provided a solution and said to use titanic_data as the name, Google was unclear and gave multiple different answers like the data is not in memory (which was an error I encountered before).

In [5]:
import pandas as pd 

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
data = titanic_data = pd.read_csv(url)

missing_values = titanic_data.isnull().sum()
print(missing_values)

data.groupby('Pclass')['Embarked'

SyntaxError: incomplete input (1332499703.py, line 9)

In [ ]:
#Chat did not understand at first, had to give another prompt. after which it picked up the missing parenthesis. Google was confused.

In [6]:
import pandas as pd 

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
data = titanic_data = pd.read_csv(url)

missing_values = titanic_data.isnull().sum()
print(missing_values)

data.group_by('Pclass')['Embarked']

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


AttributeError: 'DataFrame' object has no attribute 'group_by'

CHatGPT said to use groupby not group_by, which is correct, Google however gave multiple answers such as not not renaming it. 

In [7]:
import pandas as pd 

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
data = titanic_data = pd.read_csv(url)

missing_values = titanic_data.isnull().sum()
print(missing_values)

data.groupby('pclass')['embarked']

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


KeyError: 'pclass'

ChatGPT told me to use the upper case P instead of the lower case p where as Google wouldn't have understood the context of P. 

In [8]:
import pandas as pd 

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
data = titanic_data = pd.read_csv(url)

missing_values = titanic_data.isnull().sum()
print(missing_values)

data.groupby(Pclass)['Embarked']

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


NameError: name 'Pclass' is not defined

Chat recognised that quotation marks were supposed to be used. Google suggested I would receive an undefined value. 
9. Yes, I attended office hours

CHat history: https://chatgpt.com/share/66e3b62e-d3b4-8010-b54f-70452a618673